### PR TITLE
[feature-wip](unique-key-merge-on-write) update delete bitmap while p…

### PR DESCRIPTION
**Pull request by yixiutt**
…ublish version

1.make version publish work in version order
2.update delete bitmap while publish version, load current version rowset
primary key and search in pre rowsets
3.speed up publish version task by parallel tablet publish task

# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/task/engine_publish_version_task.cpp [hunk 1]
```diff
@@ -86,39 +115,44 @@ Status EnginePublishVersionTask::finish() {
...
-            publish_status = StorageEngine::instance()->txn_manager()->publish_txn(
-                    partition_id, tablet, transaction_id, version);
-            if (publish_status != Status::OK()) {
-                LOG(WARNING) << "failed to publish version. rowset_id=" << rowset->rowset_id()
-                             << ", tablet_id=" << tablet_info.tablet_id
-                             << ", txn_id=" << transaction_id;
-                _error_tablet_ids->push_back(tablet_info.tablet_id);
-                res = publish_status;
+            Version max_version = tablet->max_version();
+            // in uniq key model with aux index, we should see all
```
Comment (zhannngchen) [inline be/src/olap/task/engine_publish_version_task.cpp:119]: in uniq key model with merge-on-write

File: be/src/olap/task/engine_publish_version_task.cpp [hunk 2]
```diff
@@ -86,39 +115,44 @@ Status EnginePublishVersionTask::finish() {
...
+            // previous version when update delete bitmap, so add a check
+            // here and wait pre version publish or lock timeout
+            if (tablet->keys_type() == KeysType::UNIQUE_KEYS &&
+                tablet->enable_unique_key_merge_on_write() &&
+                version.first != max_version.second + 1) {
+                LOG(INFO) << "uniq key with aux index version not continuous, current max version="
```
Comment (zhannngchen) [inline be/src/olap/task/engine_publish_version_task.cpp:125]: uniq key with merge-on-write

File: be/src/olap/txn_manager.cpp [hunk 3]
```diff
@@ -308,8 +309,108 @@ Status TxnManager::publish_txn(OlapMeta* meta, TPartitionId partition_id,
...
+            // exclude last_key, last_key will be read in next batch.
+            if (num_read == batch_size && num_read != remaining) {
+                num_read -= 1;
+            }
+            for (size_t i = 0; i < num_read; i++) {
+                const Slice* key = reinterpret_cast<const Slice*>(cvb->cell_ptr(i));
+                // first check if exist in pre segment
+                bool find = _check_pk_in_pre_segments(pre_segments, *key, tablet, version);
+                if (find) {
+                    cnt++;
```
Comment (zhannngchen) [inline be/src/olap/txn_manager.cpp:374]: we should `continue` here?

File: be/src/olap/tablet.h [hunk 4]
```diff
@@ -438,6 +438,9 @@ inline void Tablet::set_cumulative_layer_point(int64_t new_point) {
 }
 
 inline bool Tablet::enable_unique_key_merge_on_write() const {
+    if (_tablet_meta == nullptr) {
```
[the remark this answers is not in the document]
Comment (yixiutt) [reply be/src/olap/tablet.h:441]: actually it's for test

File: be/src/olap/txn_manager.cpp (hunk 3 above)
Comment (yixiutt) [reply be/src/olap/txn_manager.cpp:374]: fixed

File: be/src/olap/tablet.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhannngchen) [reply be/src/olap/tablet.h:442]: ok

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Review by dataroaring — DISMISSED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**